### PR TITLE
Support bufferization of all LinalgExt ops.

### PR DESCRIPTION
**Pull request by hanhanW**
Fixes https://github.com/google/iree/issues/9004

### CODE REVIEW

**Comment by hanhanW** [comment]
I'll send out another PR to switch pipeline to use upstream comprehensive bufferization. This needs more work because we do not consider early bufferized ops in bufferizableInPlaceAnalysis. We should either

1. create another pipeline which excludes bufferization step.
2. stop bufferization in runOneShotBufferize when the ops are already bufferized.

It runs into errors in `OneShotAnalysis.cpp:152` because they are already bufferized.

@matthias-springer is there a way to fix it in runOneShotBufferize? If we can address it in bufferization pass, we don't have to create a new pipeline for this case. What's happening here is that we bufferize the kernel at an earlier stage, so there are memrefs in the IR. The IREEComprehensiveBufferize does not handle the case.

Here is the IR before IREEComprehensiveBufferize.

```mlir
module {
  func @_concatenate_cst_dispatch_1() {
    %c3 = arith.constant 3 : index
    %c2 = arith.constant 2 : index
    %c0 = arith.constant 0 : index
    %cst = arith.constant dense<0> : tensor<2x3xi32>
    %0 = bufferization.to_memref %cst : memref<2x3xi32>
    %1 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : memref<2x5xi32>
    memref.assume_alignment %1, 64 : memref<2x5xi32>
    %2 = memref.subview %1[0, 2] [2, 3] [1, 1] : memref<2x5xi32> to memref<2x3xi32, affine_map<(d0, d1) -> (d0 * 5 + d1 + 2)>>
    %workgroup_id_x = hal.interface.workgroup.id[0] : index
    %workgroup_count_x = hal.interface.workgroup.count[0] : index
    %workgroup_id_y = hal.interface.workgroup.id[1] : index
    %workgroup_count_y = hal.interface.workgroup.count[1] : index
    scf.for %arg0 = %workgroup_id_y to %c2 step %workgroup_count_y {
      %3 = affine.apply affine_map<()[s0] -> (s0 * 3)>()[%workgroup_id_x]
      %4 = affine.apply affine_map<()[s0] -> (s0 * 3)>()[%workgroup_count_x]
      scf.for %arg1 = %3 to %c3 step %4 {
        %5 = memref.subview %0[%arg0, %arg1] [1, 3] [1, 1] : memref<2x3xi32> to memref<1x3xi32, affine_map<(d0, d1)[s0] -> (d0 * 3 + s0 + d1)>>
        %6 = memref.subview %2[%arg0, %arg1] [1, 3] [1, 1] : memref<2x3xi32, affine_map<(d0, d1) -> (d0 * 5 + d1 + 2)>> to memref<1x3xi32, affine_map<(d0, d1)[s0] -> (d0 * 5 + s0 + d1)>>
        linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%5 : memref<1x3xi32, affine_map<(d0, d1)[s0] -> (d0 * 3 + s0 + d1)>>) outs(%6 : memref<1x3xi32, affine_map<(d0, d1)[s0] -> (d0 * 5 + s0 + d1)>>) attrs =  {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[1, 3]]>} {
        ^bb0(%arg2: i32, %arg3: i32):
          linalg.yield %arg2 : i32
        }
      }
    }
    return
  }
}
```

Error log:

```
mlir:BufferizationAliasInfo::bufferizeOutOfPlace(mlir::OpOperand &): Assertion `!inplaceBufferized.contains(&operand) && "OpOperand was already decided to bufferize inplace"' failed.

```

**Comment by matthias-springer** [comment]
This op is currently not supported in One-Shot Bufferize:

```
%0 = bufferization.to_memref %cst : memref<2x3xi32>
```

It is a bit tricky to support this op (and also `bufferization.to_tensor`) because they cross the memref/tensor boundary and we it is hard to track what happens to a tensor.

Ideally, you would be running One-Shot Bufferize first, and then the other bufferization later. If necessary, you can specify a filter on `BufferizationOptions` to exclude certain ops from One-Shot Bufferize. Would that be possible?

Another option would be to run One-Shot Bufferize without an analysis. This would work well in your example here, but it is problematic if you have ops that are writing to memory. E.g., a `tensor.insert` would then always result in an alloc+copy. So whether this is viable or not depends on what kind of IR you want to bufferize.

**Comment by MaheshRavishankar** [comment]
We can migrate the pass to use `IREEComprehensiveBufferize`. Not sure what issue will happen. I knew for sure we can use LinalgBufferize to get going. Do you want me to take a stab at doing that?

**Comment by MaheshRavishankar** [comment]
@matthias-springer for that assertion, we can remove the assertion and just do nothing if the operation if the operand is not contained in the `inPlaceBufferized` object?

**Comment by hanhanW** [comment]
> We can migrate the pass to use `IREEComprehensiveBufferize`. Not sure what issue will happen. I knew for sure we can use LinalgBufferize to get going. Do you want me to take a stab at doing that?

Yes, please go for it. I'll file an issue and assign to you. In the meantime, we still can review and land this PR. This PR adds support for LinalgExt ops bufferization. The lit test covers all the LinalgExt ops.

File: compiler/src/iree/compiler/Codegen/Common/test/iree_comprehensive_bufferize.mlir [hunk 1]
```diff
@@ -272,3 +272,116 @@ func @fft_tensor(%idx: index) -> (tensor<1024xf32>, tensor<1024xf32>) {
...
+// CHECK:        scf.for %[[I:.+]] = %{{.+}} to %{{.+}} step %{{.+}}
+// CHECK-DAG:      %[[SUB_UPDATE:.+]] = memref.subview %[[UPDATE]][%[[I]]]
+// CHECK-DAG:      %[[SUB_INDICES:.+]] = memref.subview %[[INDICES]][%[[I]], 0]
+// CHECK:          iree_linalg_ext.scatter
+// CHECK-SAME:     ins(%[[SUB_UPDATE]], %[[SUB_INDICES]]
+// CHECK-SAME:     outs(%[[ORIGINAL:.+]]
+
+// -----
+
+func.func @topk() {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/test/iree_comprehensive_bufferize.mlir:357]: @KoolJBlack FYI

File: compiler/src/iree/compiler/Codegen/Interfaces/BufferizationInterfaces.cpp [hunk 2]
```diff
@@ -235,8 +235,12 @@ struct LinalgExtOpInterface
                                                     OpTy> {
   bool bufferizesToMemoryRead(Operation *op, OpOperand &opOperand,
                               const AnalysisState &state) const {
-    // All operands (including outputs) may be read.
-    return true;
+    // TODO: Implement payloadUsesValueFromOperand for individual ops. There
+    // are only limit LinalgExt ops, so we hard-coded here. We don't expect
```
Comment (matthias-springer) [inline compiler/src/iree/compiler/Codegen/Interfaces/BufferizationInterfaces.cpp:239]: a limited number of

Comment (matthias-springer) [inline compiler/src/iree/compiler/Codegen/Interfaces/BufferizationInterfaces.cpp:239]: hardcode them here

**Pull request merged**